posts/introduction-to-graphql-with-apollo.md

```diff
@@ -0,0 +1,320 @@
+
+```js
+Book: {
+  ...
+  title(parent) {
+    return parent.title
+  }
+},
+```
+
+In addition, _resolvers_ define all possible operations that the server may receive: _query_, _mutation_ and _subscription_.
+
+# Query
+
+A _query_ is used **to fetch values from the GraphQL server**: it is therefore a data reading operation. The _payload_ return is in JSON format and different _queries_ can be executed in parallel. These _queries_ are defined in the schema and processed in the _resolver_.
+
+![query](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2022/08/graphql.query.png)
+
+GraphQL links the client to the server via the name of the method you aim to call (here, `books` highlighted in yellow). Once the request has been received, the server will check if it exists in the schema under the `Query` type and if it has the correct parameters and return type (here, no parameter but a list of `Book` in output).
+
+Then, the server searches in the _resolver_ the associate function under the `Query` field (only if the request is valid). Finally, the resolver returns the data and reorders it to match exactly the fields of the query, in the requested order (in our example, the resolver directly returns `bookData`, the book’s list in JSON format).
+
+# Interface
+
+Interfaces are abstract types **that allow fields to be shared between different objects**. Each object implementing an interface must at least have similar fields, and others if necessary.
+
+Let’s take the previous schema with the `Book` type. We could imagine adding a `Movie` type and a `Media` interface, so `Book` and `Movie` share common fields (`id` and `title`) through the implementation of the `Media` interface.
+
+```graphql
+interface Media {
+  id: ID!
+  title: String!
+}
+
+type Book implements Media {
+  id: ID!
+  title: String!
+  totalPages: Int
+  author: Person
+}
+
+type Movie implements Media {
+  id: ID!
+  title: String!
+  runTime: Int
+  director: Person
+}
+```
+
+Alternatively, multiple interfaces can be implemented using the following syntax:
+
+```graphql
+type MyObject implements Interface1 & Interface2 {
+  …
+}
+```
+
+Now imagine we wanted to fetch all the `Media` objects, by retrieving both common and specific fields of the objects implementing this interface.
+
+How does the server know the difference between a `Book` and a `Movie`?
+
+Firstly, the _query_ must be declared in the schema and defined in the associate _resolver_ (such as the **books’** _query_). Next, the special function `__resolveType` provided by apollo-server will be included in the resolver of the `Media` object. This function will allow us to define the type of implementation that will be returned according to the fields available in the object: in the case of `author` field is present, then a `Book` will be returned; otherwise, if the object has a `director` field, a `Film` will be returned.
```

Comment:
```suggestion
Firstly, the _query_ must be declared in the schema and defined in the associate _resolver_ (such as the **books’** _query_). Next, the special function `__resolveType` provided by apollo-server will be included in the resolver of the `Media` object. This function will allow us to define the type of implementation that will be returned according to the fields available in the object. In the case of the `author` field being present, then a `Book` will be returned. Otherwise, if the object has a `director` field, a `Film` will be returned.
```